test/dotnet-publish.Tests/PublishAppsWithToolsTests.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using FluentAssertions;
+using Microsoft.DotNet.Tools.Test.Utilities;
+using Microsoft.Extensions.DependencyModel;
+using Xunit;
+
+namespace Microsoft.DotNet.Tools.Publish.Tests
+{
+    public class PublishAppsWithToolsTests : TestBase
+    {
+        private static readonly HashSet<string> ExcludedLibraries = new HashSet<string>(StringComparer.Ordinal) {
+            "Microsoft.AspNetCore.Mvc"
+        };
+
+        [Fact]
+        public void PublishExcludesBuildDependencies()
+        {
+            var testInstance = TestAssetsManager.CreateTestInstance("AppWithDirectDependencyAndTypeBuild")
+                .WithLockFiles();
+
+            new RestoreCommand() { WorkingDirectory = testInstance.TestRoot }.Execute().Should().Pass();
+
+            var publishCommand = new PublishCommand(testInstance.TestRoot);
+            var publishResult = publishCommand.Execute();
+            publishResult.Should().Pass();
+
+            var publishDir = publishCommand.GetOutputDirectory(portable: true);
+
+            publishDir.Should().HaveFiles(new[]
+            {
+                // This one is directly referenced
+                "Microsoft.AspNetCore.Mvc.Core.dll"
+            });
+
+            // But these are brought in only by the type:build dependency, and should not be published
```

Comment:
What about package that is dependency of 'type:build' but is brought by another package as dependency too?